Return stable edge IDs in public transit routing JSON responses

mark-idleman opened the pull request:
**Purpose:** Return stable edge IDs in GH transit responses

**Usage:** Assuming the proper OSM/GTFS files for NYC specified in `config-nyc-pt.yml` are in their proper places, run `java -Xmx8g -jar web/target/graphhopper-web-1.0-SNAPSHOT.jar import config-nyc-pt.yml` to generate a GH graph for use in transit requests that has stable edge IDs encoded for each edge (thanks to the [recently merged change](https://github.com/replicahq/graphhopper/commit/4adfe2330cd8d865e9c87e4bb01ffb2b8447b69e) that adds stable IDs when using the `import` command).

Once the graph has been built, start the server with `java -Xmx8g -jar web/target/graphhopper-web-1.0-SNAPSHOT.jar server config-nyc-pt.yml`. For the current GTFS file used in the config, I know that Jan 21st, 2019 is a "valid date" for those PT lines, so set the date in GH to that date, choose start/end points, then use your browser's web dev features (eg, Chrome's "inspect" feature) to view the json response when you click to route the request. You'll see that for each leg of each trip, as well as for each trip as a whole, a list of stable edge IDs should be returned

**Changes made in this PR:**
- Added back the `reader-gtfs` module, for the sole purpose of "overriding" a small number of classes from that module to add the needed functionality. In particular, `CustomGraphHopperGtfs`, `CustomPtRouteResource`, and `CustomTripFromLabel` were added here; javadoc comments in each describe what was changed. All other classes from `reader-gtfs` come from the `1.0-pre42.3` dependency, as was the case prior to this PR
- Changed necessary dropwizard/jersey/etc web server code to point to custom classes (see changes to `GraphHopperBundle`)
- Added `replica-common` module, and moved all stable edge ID-specific classes to this module, because now both `web-bundle` and `reader-gtfs` need to access these classes, and it makes things much cleaner.
- Changed dependency chain from `web-bundle` -> `graphhopper-matrix 1.0-pre42.3` (which contains deps to all core GH modules), to `web-bundle` -> `reader-gtfs` -> `replica-common` -> `graphhopper-matrix 1.0-pre42.3`. This insures that the necessary modules we added here (`reader-gtfs` and `replica-common`) don't cause unnecessary complexity WRT transitive dependencies. Now, each pomfile we have has as few deps as possible, and the dep chain is very explicit and easy to follow

mark-idleman committed:
draft: add stable edge IDs in transit responses without need for dep hell

mark-idleman committed:
remove files from web-bundle that were moved to replica-common

mark-idleman committed:
Finally fix jetty injection issue by creating custom gh gtfs class

kaelgreco commented on config-nyc-pt.yml:
```diff
@@ -0,0 +1,19 @@
+graphhopper:
```
what's this file used for?

mark-idleman replied:
This is the GH config file, used to specify settings for GH/location of OSM+GTFS files/which vehicle profiles we want. The way GH knows that it should build graphs for transit requests and not auto/bike/ped is whether or not the `gtfs.file` line is included in the config (and in this case, you can't include the usual vehicles that our standard configs use), so I made a special config for the NYC region for testing here.

In a productionized setting, we'll want to come up with templating for configs like this (you can provide each of these lines as command line args, so we can make a general PT config that we inject GTFS paths/etc into)

kaelgreco commented on reader-gtfs/src/main/java/com/graphhopper/reader/gtfs/CustomGraphHopperGtfs.java:
```diff
@@ -0,0 +1,230 @@
+/*
```
This is just pulled from gh, yeah? did you need to modify anything?

mark-idleman replied:
Essentially, yes, with minor updates. Added javadoc to this class to specify what was changed/what wasn't

kaelgreco reviewed (commented):
I'm so excited about this. it changes everything.
Mostly looks good to me. Want to show me what sections are new code (i.e. not copied over from gh).

Also, where is the route id constructed? we do some funny things with r5 transit route ids

mark-idleman committed:
add javadocs + comments to all custom classes; clean up custom GH gtfs class

mark-idleman commented:
> Want to show me what sections are new code (i.e. not copied over from gh).

Added a description to the PR outlining some of this, and added javadocs/inline comments to the classes I modified specifying what was copied/what was modified for the large classes I added/messed with in this PR.

> Also, where is the route id constructed? 

Good question, not sure! I'll poke around the codebase to see where that happens, it wasn't something I touched when making these changes

mark-idleman commented:
After learning that this approach doesn't make sense in [this discussion](https://replicahq.slack.com/archives/CK358RL0Z/p1596712824378500), I'm closing this in favor of [this new approach](https://github.com/replicahq/graphhopper/pull/23), which essentially runs a more-efficient GTFS link mapper during GH's graph build process, and stores/returns the resulting mapped stabled edge IDs at query time.
